[JSC] TypedArray iteration does not need to get "length"

Constellation opened the pull request:
#### af1cc8c8b4cce68e686412df06a6646b76b96be3
<pre>
[JSC] TypedArray iteration does not need to get &quot;length&quot;
<a href="https://bugs.webkit.org/show_bug.cgi?id=243581">https://bugs.webkit.org/show_bug.cgi?id=243581</a>

Reviewed by Alexey Shvayka.

ArrayIterator#next spec says that if the array is TypedArray, we do not need to look up &quot;length&quot;,
and we can directly get TypedArray&apos;s internal length. This means that &quot;length&quot; property of instance,
Uint8Array.prototype, and TypedArray.prototype are unrelated to iterator protocol. This makes iterator
protocol guarantee simplified.

This patch applies this change: we no longer ensure &quot;length&quot; validity. We also adjust ArrayIterator#next&apos;s
slow path implementation to the spec.

[1]: <a href="https://tc39.es/ecma262/#sec-createarrayiterator">https://tc39.es/ecma262/#sec-createarrayiterator</a>

* JSTests/stress/typed-array-from-custom-length.js:
* Source/JavaScriptCore/builtins/ArrayIteratorPrototype.js:
(linkTimeConstant.arrayIteratorNextHelper):
* Source/JavaScriptCore/runtime/JSArrayBufferView.cpp:
(JSC::JSArrayBufferView::isIteratorProtocolFastAndNonObservable):
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::installTypedArrayIteratorProtocolWatchpoint):
(JSC::JSGlobalObject::installTypedArrayPrototypeIteratorProtocolWatchpoint):
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
(JSC::JSGlobalObject::typedArrayPrototypeLengthAbsenceWatchpoint): Deleted.
* Source/JavaScriptCore/runtime/JSTypedArrayViewPrototype.cpp:
(JSC::JSTypedArrayViewPrototype::finishCreation):

Canonical link: <a href="https://commits.webkit.org/253153@main">https://commits.webkit.org/253153@main</a>
</pre>

shvaikalesh reviewed (approved):
Lovely, r=me.

I wish we could export `JSTests/stress/typed-array-from-custom-length.js` and `JSTests/stress/detached-typed-array-iteration.js` to test262 suite some day.

Constellation committed:
[JSC] TypedArray iteration does not need to get "length"
https://bugs.webkit.org/show_bug.cgi?id=243581

Reviewed by Alexey Shvayka.

ArrayIterator#next spec says that if the array is TypedArray, we do not need to look up "length",
and we can directly get TypedArray's internal length. This means that "length" property of instance,
Uint8Array.prototype, and TypedArray.prototype are unrelated to iterator protocol. This makes iterator
protocol guarantee simplified.

This patch applies this change: we no longer ensure "length" validity. We also adjust ArrayIterator#next's
slow path implementation to the spec.

[1]: https://tc39.es/ecma262/#sec-createarrayiterator

* JSTests/stress/typed-array-from-custom-length.js:
* Source/JavaScriptCore/builtins/ArrayIteratorPrototype.js:
(linkTimeConstant.arrayIteratorNextHelper):
* Source/JavaScriptCore/runtime/JSArrayBufferView.cpp:
(JSC::JSArrayBufferView::isIteratorProtocolFastAndNonObservable):
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::installTypedArrayIteratorProtocolWatchpoint):
(JSC::JSGlobalObject::installTypedArrayPrototypeIteratorProtocolWatchpoint):
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
(JSC::JSGlobalObject::typedArrayPrototypeLengthAbsenceWatchpoint): Deleted.
* Source/JavaScriptCore/runtime/JSTypedArrayViewPrototype.cpp:
(JSC::JSTypedArrayViewPrototype::finishCreation):

Canonical link: https://commits.webkit.org/253153@main
